### ISSUE TITLE
Does not properly escape special characters such as backslashes in the STATUS replies

### ISSUE THREAD

**Issue by malteclasen** [hmailserver/hmailserver#85]
When using a baskslash as hierarchy separator, hMailServer returns this backslash unescaped in STATUS replies. This causes issues with some email clients such as Trojita, see https://bugs.kde.org/show_bug.cgi?id=344605 , especially Jan Kundrát's comment:

>See definition of "quoted" at http://tools.ietf.org/html/rfc3501#page-88). Given that the error message is about STATUS, and each STATUS is preceded by a LIST, this means that the server very likely manages to send LIST correctly, yet fails on STATUS.

**Comment by malteclasen** [comment]
Jan Kundrát had a closer look at the hMailServer source, and apparently there's no escaping in the STATUS code ( https://bugs.kde.org/show_bug.cgi?id=344605#c6 ).

**Comment by jktjkt** [comment]
Yeah, seems that IMAPFolder::EscapeFolderString(...) should be added to https://github.com/hmailserver/hmailserver/blob/master/hmailserver/source/Server/IMAP/IMAPCommandStatus.cpp#L58 .

**Comment by martinknafve** [comment]
This will be fixed in 5.7.

**Comment by martinknafve** [comment]
This will be fixed in 5.6.2 as well.

Martin

----- Ursprungligt meddelande -----
Från: "Malte Clasen" <notifications@github.com>
Skickat: ‎2015-‎03-‎02 21:40
Till: "hmailserver/hmailserver" <hmailserver@noreply.github.com>
Ämne: Re: [hmailserver] Does not properly escape special characters such asbackslashes in the STATUS replies (#85)

Jan Kundrát had a closer look at the hMailServer source, and apparently there's no escaping in the STATUS code ( https://bugs.kde.org/show_bug.cgi?id=344605#c6 ).

**Comment by RvdHout** [comment]
@martinknafve did this ever got fixed?

Is it fixed by just the addition of IMAPFolder::EscapeFolderString(...) mentioned by @jktjkt that should be added? 

Something like:

```
      String sFolderName = pParser->GetParamValue(pArgument, 0);
      String sFlags = pParser->GetParamValue(pArgument, 1);

      // We cannot send " or \ directly in the response.
      // We must escape those:
      IMAPFolder::EscapeFolderString(sFolderName);

      std::shared_ptr<IMAPFolder> pTheFolder = pConnection->GetFolderByFullPath(sFolderName);
      if (!pTheFolder)
         return IMAPResult(IMAPResult::ResultBad, "Folder could not be found.");

      // Check if ther user has access to read this folder.
      if (!pConnection->CheckPermission(pTheFolder, ACLPermission::PermissionRead))
         return IMAPResult(IMAPResult::ResultBad, "ACL: Read permission denied (Required for STATUS command).");

      // Check if ther user has access to read this folder.
      if (!pConnection->CheckPermission(pTheFolder, ACLPermission::PermissionRead))
         return IMAPResult(IMAPResult::ResultBad, "ACL: Read permission denied.");

      std::shared_ptr<Messages> pMessages = pTheFolder->GetMessages();
      
      String sResponse = "";

      bool bAddSpace = false;

      sResponse = "* STATUS \"" + sFolderName + "\" ";
```

**Comment by alecpl** [comment]
I'm not hMS developer, but the answer is obvious. Yes.

If you look at https://github.com/hmailserver/hmailserver/blob/12d4d4c9f73ac1534b8dcb661799ad883871fb07/hmailserver/source/Server/IMAP/FolderListCreator.cpp#L132 there's also `AdjustCaseToClientCase_` call that may or may not be needed.

If you read RFC 3501 you'll know that some "strings" do not require quotes and some others use even different syntax (literal). https://datatracker.ietf.org/doc/html/rfc3501#section-4.3

**Comment by martinknafve** [comment]
Fix merged.
